fabric-rendering-v1: Custom Armor Model & Texture

shedaniel opened the pull request:
**This PR mainly adds two classes:**
**ArmorModelRegistry** -> A class for registering custom armor models for items, to be provided by a contextual provider.
**ArmorTextureRegistry** -> A class for registering custom armor textures for items, to be provided by a contextual provider. Since armor textures identifier in vanilla is hardcoded to be in the `minecraft` namespace, this registry can also be used to use a custom namespace if desired.

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
Moving testmod id to rendering-v1

Signed-off-by: shedaniel <daniel@shedaniel.me>

Moving testmod to rendering-v1

Signed-off-by: shedaniel <daniel@shedaniel.me>

Reviews

Signed-off-by: shedaniel <daniel@shedaniel.me>

prefix the extensions with armor

Signed-off-by: shedaniel <daniel@shedaniel.me>

change name

Signed-off-by: shedaniel <daniel@shedaniel.me>

drop custom

Signed-off-by: shedaniel <daniel@shedaniel.me>

thing

Signed-off-by: shedaniel <daniel@shedaniel.me>

javadocs

Signed-off-by: shedaniel <daniel@shedaniel.me>

fix imports

Signed-off-by: shedaniel <daniel@shedaniel.me>

forgot to do asItem

Signed-off-by: shedaniel <daniel@shedaniel.me>

add null checks and convert to ItemConvertible

Signed-off-by: shedaniel <daniel@shedaniel.me>

fix license

Signed-off-by: shedaniel <daniel@shedaniel.me>

did thing

Signed-off-by: shedaniel <daniel@shedaniel.me>

it now compiles

Signed-off-by: shedaniel <daniel@shedaniel.me>

change to a registry

Signed-off-by: shedaniel <daniel@shedaniel.me>

add @Unique

Signed-off-by: shedaniel <daniel@shedaniel.me>

migrate to fabric-item-api-v1

Signed-off-by: shedaniel <daniel@shedaniel.me>

did some renaming and improvements

Signed-off-by: shedaniel <daniel@shedaniel.me>

don't need that

Signed-off-by: shedaniel <daniel@shedaniel.me>

armor

Signed-off-by: shedaniel <daniel@shedaniel.me>

i509VCB commented on fabric-armor-api-v1/src/main/java/net/fabricmc/fabric/api/client/armor/v1/CustomTexturedArmor.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Implement this interface on an Item which you would like to have a custom textured armor model.
+ */
+public interface CustomTexturedArmor {
```
Honestly I think just implementing the interface and using the item's registry namespace and path would probably suffice over a whole method.

shedaniel replied:
Well possibly, I may default the return value to the item registry namespace? Since I really want the customizability here without limiting due to the item registry namespace.

It will be useful for items that use the same texture (and maybe using the different UV).

i509VCB replied:
I could see a reason making the selected texture stack aware and slightly more contextual.

i509VCB commented:
I'd consider possibly another approach to the properly identified item texture, I have one version that uses modify arg instead of the identifier.

https://github.com/i509VCB/Fabric-Junkkyard/blob/master/src/main/java/me/i509/junkkyard/armor/mixin/ArmorFeatureRendererMixin.java

i509VCB commented on fabric-armor-api-v1/src/main/java/net/fabricmc/fabric/mixin/armor/client/ArmorFeatureRendererMixin.java:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	@Inject(method = "getArmor", at = @At("RETURN"), cancellable = true)
+	private void getArmor(EquipmentSlot slot, CallbackInfoReturnable<BipedEntityModel<LivingEntity>> cir) {
```
Injection method doesn't describe very well what this does. Maybe `selectArmorModel`

JamiesWhiteShirt commented:
An alternative for textures on armors without custom models it to pre-register the texture. Using an accessor for the map that is already there, it just works ™️, no interface implementation required.

https://github.com/JamiesWhiteShirt/hard-hat-steve/blob/d18ba349d00a34f876a7b21f6e7eaee8f56017c2/src/main/java/com/jamieswhiteshirt/hardhatsteve/client/HardHatSteveClient.java#L23-L36

i509VCB commented:
Accessors should be impl detail. We can't default to fixing namespace automatically or all mods that add armor break

shedaniel commented:
If you use the accessor only method, it will limit the API since it will be determined by the armor material name. Therefore it will not support different textures for the same armor material.

I'll take a look at the ModifyArg approach.

shedaniel commented:
Question: How compatible is @ModifyArg if two mods edit the same thing?

i509VCB commented:
if two mods an argument at the same index both fail

liach commented:
I feel this is better suited for item's client api than a new separate armor api

i509VCB commented:
Yeah I could see that an armor micro module isn't 100% nessecary.

shedaniel commented:
I'll move this to fabric-item-api-v1

shedaniel commented:
I've moved this to `fabric-item-api-v1`

modmuss50 commented:
Would the rendering module be a better fit for this?

shedaniel commented:
the only other thing in fabric-item-api-v1 is the item tooltip, I don't know

modmuss50 commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/CustomModeledArmor.java:
```diff
@@ -0,0 +1,44 @@
+import net.fabricmc.api.Environment;
+
+/**
+ * Implement this interface on an {@link net.minecraft.item.ArmorItem} which you would like to have a custom armor model.
```
I dont think we should really have what should be client only interfaces like this, Vanilla is moving away from this (think render layers) a registry would be better IMO.

shedaniel replied:
think it's easier to have an interface, for example: `BlockEntityClientSerializable`, you don't want to have an registry for such a thing. since the armor models and textures might depend on the item itself, I think it is more simple and easier to have it be an interface.

modmuss50 replied:
BlockEntityClientSerializable is for the server tho. A vanilla example of this is the BlockEntityRenderers, there is a clear speration between client and server code, without relaying on the annotations. A registry might make it easier for a mod to change the model of lets say the vanilla armor without requiring a mixin.

shedaniel replied:
hmm true, I agree that it will be easier to change the vanilla armor that way. I will change this into a registry

shedaniel commented:
this is now a registry

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/CustomModeledArmorRegistry.java:
```diff
@@ -0,0 +1,30 @@
+
+@Environment(EnvType.CLIENT)
+public interface CustomModeledArmorRegistry extends CustomModeledArmorProvider {
+	CustomModeledArmorRegistry INSTANCE = CustomModeledArmorRegistryImpl.INSTANCE;
```
I personally want to move away from the `instance` design of this. Personally I think we should make this entirely static so we chop off the `INSTANCE` call.

Also why is this itself an armor provider. That should really be impl detail.

shedaniel replied:
> Also why is this itself an armor provider. That should really be impl detail.

why not, it's just duplicate methods, but if we are moving away from `INSTANCE` this can go

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/CustomTexturedArmorRegistry.java:
```diff
@@ -0,0 +1,33 @@
+import net.fabricmc.fabric.impl.client.item.CustomTexturedArmorRegistryImpl;
+
+/**
+ * Implement this interface on an {@link net.minecraft.item.ArmorItem} which you would like to have a custom armor model.
```
Why would you implement an interface meant for impl

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/CustomTexturedArmorRegistry.java:
```diff
@@ -0,0 +1,33 @@
+ */
+@Environment(EnvType.CLIENT)
+public interface CustomTexturedArmorRegistry extends CustomTexturedArmorProvider {
+	CustomTexturedArmorRegistry INSTANCE = CustomTexturedArmorRegistryImpl.INSTANCE;
```
I think a static method would suffice over the instance.

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/impl/client/item/CustomTexturedArmorRegistryImpl.java:
```diff
@@ -0,0 +1,57 @@
+
+public final class CustomTexturedArmorRegistryImpl implements CustomTexturedArmorRegistry {
+	public static final CustomTexturedArmorRegistry INSTANCE = new CustomTexturedArmorRegistryImpl();
+	private final Map<Item, CustomTexturedArmorProvider> providerMap = new Reference2ObjectOpenHashMap<>();
```
Maybe mixin to item and store the provider there to reduce the `map.get` calls

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/impl/client/item/CustomModeledArmorRegistryImpl.java:
```diff
@@ -0,0 +1,58 @@
+
+public final class CustomModeledArmorRegistryImpl implements CustomModeledArmorRegistry {
+	public static final CustomModeledArmorRegistry INSTANCE = new CustomModeledArmorRegistryImpl();
+	private final Map<Item, CustomModeledArmorProvider> providerMap = new Reference2ObjectOpenHashMap<>();
```
I think we should mixin to item and store the model provided on items to reduce `map.get` calls

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/impl/client/item/CustomTexturedArmorRegistryImpl.java:
```diff
@@ -0,0 +1,57 @@
+	@Override
+	public void registerCustomArmorTexture(CustomTexturedArmorProvider provider, Item... items) {
+		for (Item item : items) {
+			providerMap.put(item, provider);
```
What about duplicates, how should we handle those?

shedaniel replied:
we don't, the one registered last gets it

i509VCB replied:
That somewhat seems like a load order battle

shedaniel replied:
you can't have two textures at the same time, also that's rare since usually you only edit your own item. If two mods edit the same vanilla item, you can't guarantee compatibility between them at all.

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/CustomModeledArmorRegistry.java:
```diff
@@ -0,0 +1,82 @@
+	 * @param provider the provider for the model
+	 * @param items    the items to be registered for
+	 */
+	public static void registerCustomArmorModel(/* @Nullable */ Provider provider, Item... items) {
```
I don't really know why you would want a null provider for this.

shedaniel replied:
to maybe unregister? I don't see a point to limit the api

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/CustomModeledArmorRegistry.java:
```diff
@@ -0,0 +1,82 @@
+	 * @param provider the provider for the model
+	 * @param items    the items to be registered for
+	 */
+	public static void registerCustomArmorModel(/* @Nullable */ Provider provider, Item... items) {
```
I think we should use `ItemConvertible` here in case someone decides to make a literal armor block so you don't need an `asItem` call.

i509VCB replied:
I think you should null check the items varargs.

shedaniel replied:
True

liach replied:
the null check can implicitly happen when iterating

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/CustomTexturedArmorRegistry.java:
```diff
@@ -0,0 +1,86 @@
+	 * @param provider the provider for the texture
+	 * @param items    the items to be registered for
+	 */
+	public static void registerCustomArmorTexture(/* @Nullable */ Provider provider, Item... items) {
```
Like the other api class, same questions

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/CustomModeledArmorRegistry.java:
```diff
@@ -0,0 +1,82 @@
+	 * @param provider the provider for the model
+	 * @param items    the items to be registered for
+	 */
+	public static void registerCustomArmorModel(/* @Nullable */ Provider provider, Item... items) {
```
Would there be a specific reason you'd want to register the same armor model to several items at once?

shedaniel replied:
It's not the same model, it's a model provider.

i509VCB commented:
Also the question of duplicates. What if two mods want to set the model for the same item

shedaniel commented:
The answer to https://github.com/FabricMC/fabric/pull/963#discussion_r465267686 should also apply to models.

i509VCB commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/CustomModeledArmorRegistry.java:
```diff
@@ -0,0 +1,83 @@
+		 * @param defaultModel The default model that vanilla provides
+		 * @return The model of the armor piece. Should never be null.
+		 */
+		BipedEntityModel<LivingEntity> getCustomArmorModel(LivingEntity entity, ItemStack stack, EquipmentSlot slot, BipedEntityModel<LivingEntity> defaultModel);
```
Mark method or interface as client only

i509VCB commented:
atleast codewise things look fine

Juuxel commented:
`CustomModeledArmorRegistry` and `CustomTexturedArmorRegistry` sound a bit weird to me, maybe `ArmorModelRegistry` and `ArmorTextureRegistry` since that's what they register?

shedaniel commented:
true, I like that more, ill drop the custom prefix

Player3324 commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorModelRegistry.java:
```diff
@@ -0,0 +1,92 @@
+ * <p>
+ * This can be used to replace existing vanilla armor models conditionally, however each {@link net.minecraft.item.Item}
+ * instance can only allow one {@link Provider}.
+ * </p>
```
No optional closing tags, opening tag in the same line without space please.

Player3324 commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorModelRegistry.java:
```diff
@@ -0,0 +1,92 @@
+	 * @param provider the provider for the model
+	 * @param items    the items to be registered for
+	 */
+	public static void register(/* @Nullable */ Provider provider, ItemConvertible... items) {
```
Not sure about using ItemConvertible over Item. Blocks are unusual for armor and for the rare exception the caller can convert easily. Item may be less confusing. The implementation mandates ArmorItem, so maybe we should use/require that in the API as well.

liach replied:
Hmm, can't carved pumpkins be used as armor

Juuxel replied:
Carved pumpkins aren't armor but `BlockItem`s and they use their item model instead of an armor model (this behavior is probably specific to the head slot).

Player3324 replied:
The instrumented code starts with `if (itemStack.getItem() instanceof ArmorItem) {`

Player3324 replied:
More precisely, the pumpkin or worn heads use separate feature renderers that aren't covered by this PR.

Player3324 commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorModelRegistry.java:
```diff
@@ -0,0 +1,92 @@
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.item.v1;
```
We concluded in Discord to put this in the rendering module.

Player3324 commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorModelRegistry.java:
```diff
@@ -0,0 +1,92 @@
+	 * @param items    the items to be registered for
+	 */
+	public static void register(/* @Nullable */ Provider provider, Iterable<ItemConvertible> items) {
+		ArmorModelRegistryImpl.register(provider, Objects.requireNonNull(items));
```
We usually check preconditions first, then blank line, then implementation.

i509VCB replied:
Not documented in conventions. Will add that to doc for future.

Player3324 commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorTextureRegistry.java:
```diff
@@ -0,0 +1,100 @@
+	 * @param items    the items to be registered for
+	 */
+	public static void register(/* @Nullable */ Provider provider, Iterable<ItemConvertible> items) {
+		ArmorTextureRegistryImpl.register(provider, Objects.requireNonNull(items));
```
Objects.requireNonNull use is inconsistent wrt relying on later rejection or not, in this case the implementation would crash already.

Player3324 commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorTextureRegistry.java:
```diff
@@ -0,0 +1,100 @@
+ * <p>
+ * Since armor textures identifier in vanilla is hardcoded to be in the {@code minecraft} namespace, this registry can also be
+ * used to use a custom namespace if desired.
+ * </p>
```
This needs more documentation on how models and textures relate to each other.

The suggestion around replacing vanilla armor should be more like a warning along the lines of "While this can be used to replace vanilla armor, there can be only one replacement at a time, causing potential conflicts with other mods. Consider using a separate item instead." 

Armor models and texture are so tightly connected that a single ArmorRenderingRegistry class may be more appropriate.

IMO we should also offer a wider hook that effectively replaces all of `renderArmor()` with a handler for more invasive customization.

shedaniel replied:
> IMO we should also offer a wider hook that effectively replaces all of renderArmor() with a handler for more invasive customization.

The BipedEntityModel _should be_ possible to be entirely customized, you can completely rip out BipedEntityModel#render.

Player3324 commented on fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorTextureRegistry.java:
```diff
@@ -0,0 +1,100 @@
+		 * @param entity      The entity equipping the armor
+		 * @param stack       The item stack of the armor
+		 * @param slot        The slot which the armor is in
+		 * @param secondLayer Whether the texture should use the second layer, only true when the {@code slot} is {@link EquipmentSlot#LEGS}
```
I don't think we should pass secondLayer like that while also supplying the slot. It is just a weird artefact of MC's implementation. Likewise suffix should probably be replaced with boolean overlay to denote the 2nd render pass without dye.

We could instead supply the full vanilla path in the `textures/models/armor/[..]_layer_[..].png` form to make this easy for namespace replacement use cases, the proposed ArmorModelRegistry already does something similar with defaultModel.

Player3324 reviewed (commented):
I have a few ideas on how to improve this API besides a couple minor nitpicks, please see the comments inline.

shedaniel committed:
add license to CustomArmorTests

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
Add @Nullable annotations and fix compile

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel commented:
This has been updated to target 1.16.2/3, moved to rendering-v1 and added @Nullable annotations

shedaniel committed:
javadoc

Signed-off-by: shedaniel <daniel@shedaniel.me>

NinjaPhenix commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/ArmorRenderingRegistry.java:
```diff
@@ -0,0 +1,149 @@
+	 * @param defaultTexture The default vanilla texture identifier
+	 * @return the custom armor texture identifier, return null to use the vanilla ones. Defaulted to the item's registry id.
+	 */
+	/* @Nullable */
```
Missed annotation comment?

NinjaPhenix commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/ArmorProviderExtensions.java:
```diff
@@ -0,0 +1,31 @@
+import net.fabricmc.fabric.api.client.rendering.v1.ArmorRenderingRegistry;
+
+public interface ArmorProviderExtensions {
+	/* @Nullable */
```
Two missed annotation comments

Player3324 reviewed (approved):
Good now, I think. Thanks!

shedaniel committed:
Fix reviews

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
Merge remote-tracking branch 'fabric/1.16' into armor

# Conflicts:
#	fabric-rendering-v1/build.gradle

modmuss50 reviewed (approved):
Ready for last call, thanks 👍

JamiesWhiteShirt commented on fabric-rendering-v1/src/testmod/java/net/fabricmc/fabric/test/rendering/client/CustomArmorTests.java:
```diff
@@ -0,0 +1,78 @@
+import net.fabricmc.fabric.api.client.rendering.v1.ArmorRenderingRegistry;
+
+@Environment(EnvType.CLIENT)
+public class CustomArmorTests implements ClientModInitializer {
```
This (client) mod initializer registers items, meaning that it is not dedicated server compatible. It would be very simple to split this into two initializers in case someone would like to test this with a server.

modmuss50 replied:
Good point, IMO this is a blocker for merging.

JamiesWhiteShirt commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinArmorFeatureRenderer.java:
```diff
@@ -0,0 +1,103 @@
+
+	@Inject(method = "getArmorTexture", at = @At(value = "INVOKE", target = "Ljava/util/Map;computeIfAbsent(Ljava/lang/Object;Ljava/util/function/Function;)Ljava/lang/Object;"), cancellable = true, locals = LocalCapture.CAPTURE_FAILHARD)
+	private void getArmorTexture(ArmorItem armorItem, boolean secondLayer, /* @Nullable */ String suffix, CallbackInfoReturnable<Identifier> cir, String vanillaIdentifier) {
+		String model = ArmorRenderingRegistry.getArmorTexture(storedEntity, storedEntity.getEquippedStack(storedSlot), storedSlot, vanillaIdentifier);
+
+		if (!Objects.equals(model, vanillaIdentifier)) {
+			cir.setReturnValue(ARMOR_TEXTURE_CACHE.computeIfAbsent(model, Identifier::new));
```
```suggestion
		String texture = ArmorRenderingRegistry.getArmorTexture(storedEntity, storedEntity.getEquippedStack(storedSlot), storedSlot, vanillaIdentifier);

		if (!Objects.equals(texture, vanillaIdentifier)) {
			cir.setReturnValue(ARMOR_TEXTURE_CACHE.computeIfAbsent(texture, Identifier::new));
```

JamiesWhiteShirt reviewed (changes requested):
I believe that the majority of developers will create mods using default armor models with custom textures (such as Hard Hat Steve). The path of least resistance is only putting the textures in the right namespace. This API could be made more convenient for that use case.

For this, I suggest adding a utility method for registering a "simple" armor texture in the right namespace. Internally, it could register a TextureProvider that simply sets a specified namespace on the default texture.

shedaniel committed:
Update fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinArmorFeatureRenderer.java

Co-authored-by: Erlend Åmdal <erlend@aamdal.com>

modmuss50 commented:
I do like the idea that @JamiesWhiteShirt suggested of a simple api for most people.

The test mods need fixing as well. I will leave it last call and it can be merged shortly after fixing.

modmuss50 committed:
Add registerSimpleTexture
Pass through secondLayer and suffix
Use Identifier's over strings
Fix the test mod

modmuss50 commented:
I have updated this PR with some minor (ish) changes.

modmuss50 committed:
license fix

Player3324 reviewed (approved):
Still good to me!
